Use `path = [bala]` for `rust-bitcoin`'s workspace members dependencies. remove `[patch.crates-io.balab]`

eval-exec opened the pull request:
This PR want to:

1. make all workspace members use `workspace = true` syntax to import dependencies.
2. use `path = [balabala]` to define dependencies, instead of useing `[patch.crates-io.balabala]` , fix: https://github.com/rust-bitcoin/rust-bitcoin/issues/4283

apoelstra commented:
Doesn't work with our MSRV, sorry. Otherwise we would've done this a while ago. I agree that the current situation is pretty annoying.

eval-exec commented:
Hello @apoelstra , Could you guide me where is the file: `./maintainer-tools/ci/run_task.sh` I can't found it. And I wonder why the CI (`MSRV toolchain (minimal)` and `MSRV toolchain (recent)`) failed.

apoelstra commented:
Just try to build the project with 1.63.0, like `cargo +1.63.0 test`.

But the relevant files are in https://github.com/rust-bitcoin/rust-bitcoin-maintainer-tools

eval-exec commented:
> cargo +1.63.0 test

```bash
❯ cargo +1.63.0 test
error: failed to load manifest for workspace member `/home/exec/Projects/github.com/rust-bitcoin/rust-bitcoin/base58`

Caused by:
  failed to parse manifest at `/home/exec/Projects/github.com/rust-bitcoin/rust-bitcoin/base58/Cargo.toml`

Caused by:
  feature `workspace-inheritance` is required

  The package requires the Cargo feature called `workspace-inheritance`, but that feature is not stabilized in this version of Cargo (1.63.0 (fd9c4297c 2022-07-01)).
  Consider trying a newer version of Cargo (this may require the nightly release).
  See https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#workspace-inheritance for more information about the status of this feature.
```

OK, the MSRV 1.63.0 not support [Workspace Inheritance](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#workspace-inheritance), So I'll remove `workspace = true` related stuff, but I suggest we should still use `path = [balabala]`, and remove `[patch.crates-io.balaba]`. What do you think?

Kixunil reviewed (changes requested):
Sadly, I don't believe this is possible. The reason is `secp256k1` depends on `bitcoin-hashes` via crates.io so doing this will cause it to pull the version from there rather than here. Since we removed `ThirtyTwoByteHash` this is not longer as bad as before but it still has the problem that it'll stop us from testing the chagnes to `hashes` with `secp256k1`.

However, if I understand it correctly, `[patch]` for `hashes` can still be there, so I request that you keep that one.

eval-exec commented:
> However, if I understand it correctly, `[patch]` for `hashes` can still be there, so I request that you keep that one.

Thank you, I added a comment to keep `[patch]` for `[hashes]`.

apoelstra commented:
Thanks for your continued work on this. I think it's probably fine to remove the patch section for everything but `hashes` (or at least, it will become increasingly fine as we remove `hashes` from our public APIs of downstream crates).

Can you squash all these commits together, and add a comment next to the `hashes` patch explaining why it's necessary when all the other dependencies use `path`.

eval-exec commented:
> Can you squash all these commits together, and add a comment next to the `hashes` patch explaining why it's necessary when all the other dependencies use `path`.

Thank you, I squashed and added comments for `[patch.crates-io.hashes]`.

Kixunil reviewed (approved):
ACK 9a572dabdeb077f96b2ab66be1a80fcec3e805e3

apoelstra reviewed (approved):
ACK 9a572dabdeb077f96b2ab66be1a80fcec3e805e3; successfully ran local tests

tcharding commented:
This breaks `cargo publish`. What was the motivation for this PR? I don't see how #4283 led to this PR.

apoelstra commented:
@tcharding without this PR it's a bit annoying to directly depend on the repo. You basically have to copy/paste all the `[patch]` lines from this repo into your own repo.

I don't think this is causing `cargo publish` to fail though -- I think it's the lack of version.

The exact error is

```
error: all dependencies must have a version specified when publishing.
dependency `bitcoin-internals` does not specify a version
Note: The published dependency will use the version from crates.io,
the `path` specification will be removed from the dependency declaration.
```

The `path` thing seems to be a "Note:" while the error is about a missing version.
